src/main/java/com/google/impactdashboard/server/api_utilities/IamBindingRetriever.java

```diff
@@ -0,0 +1,121 @@
+        .setApplicationName("Recommendation Impact Dashboard")
+        .build();
+    return new IamBindingRetriever(new HashMap<>(), iamService);
+  }
+
+  /**
+   * Takes audit logs and uses the IAM API with the bindings from the audit logs to
+   * calculate the IAMBindingsNumber for Each log.
+   * @param logEntries List of audit logs that set IAM policy
+   * @return the Database entries of IAMBindings from the audit logs
+   */
+  public List<IAMBindingDatabaseEntry> listIAMBindingData(Collection<LogEntry> logEntries,
+                                      String projectId, String projectName,
+                                      String projectNumber){
+    Map<Timestamp, AuditLog> timeToAuditLogMap = logEntries.stream().map(log -> {
+          AuditLog auditLog;
+          try {
+            auditLog = AuditLog.parseFrom(log.getProtoPayload().getValue());
+          } catch (InvalidProtocolBufferException e) {
+            throw new RuntimeException("Invalid Protocol Buffer used");
+          }
+          return new SimpleImmutableEntry<>(log.getTimestamp(), auditLog);
+        }).collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+
+    return timeToAuditLogMap.entrySet().stream().map(entry -> {
+      Map<String, Integer> membersForRoles = getMembersForRoles(entry.getValue().getResponse()
+          .getFieldsMap().get("bindings").getListValue().getValuesList());
+      try {
+        return IAMBindingDatabaseEntry.create(projectId, projectName, projectNumber, entry.getKey().getSeconds(),
+            getIamBindings(membersForRoles));
+      } catch (IOException e) {
+        throw new RuntimeException("Error in getting IAM Roles");
+      }
+    }).collect(Collectors.toList());
+  }
+
+  /**
+   * Takes in a map of protobuf value to value and updates the mapRoleToNumberOfMembers field with the new
+   * role and the number of members in the role.
+   * @param bindings protoBuf map for a binding map from AuditLogs
+   */
+  private Map<String, Integer> getMembersForRoles(List<Value> bindings) {
+    Map<String, Integer> membersforRoles = new HashMap<>();
+    bindings.forEach(bindingValue -> {
+      Map<String, Value> bindingMap = bindingValue.getStructValue().getFieldsMap();
+      membersforRoles.put(bindingMap.get("role").getStringValue(),
+          bindingMap.get("members").getListValue().getValuesList().size());
+    });
+    return membersforRoles;
+  }
+
+  /**
+   * Calls IAM API to get all roles for a project and then calculate the IAM bindings for the given map.
+   * @param membersForRoles map of membersPerRole to calculate the number of total bindings.
+   * @return Total number of IAMBindings for the given map
+   */
+  private int getIamBindings(Map<String, Integer> membersForRoles) throws IOException {
+    int iamBindings;
+    List<Role> roles = iamService.roles().list().execute().getRoles();
+    iamBindings = roles.stream().filter(role -> membersForRoles.containsKey(role.getName()))
```

Comment:
i think this can be just be "int iamBindings = ...", then you can delete the "int iamBindings;" above that doesn't have a value.